### PR TITLE
Add MapFind class to facilitate search in a map without iterators

### PR DESCRIPTION
**Pull request by nonifier**

### CODE REVIEW

File: map_find.h [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            res = boost::make_optional(cref_val);
+        } else {
+            res = boost::none;
+        }
+        return *this;
+    }
+
+private:
+    Map & map;
+    boost::optional<ConstRefValueType> res;
```
Comment (TeXitoi) [inline map_find.h]: a boost::optional on a std::reference_wrapper is a bit useless, you can just use a pointer.

File: tests/map_find_test.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+#include <boost/test/unit_test.hpp>
+#include <map>
+#include <unordered_map>
+#include <string>
+
+using namespace navitia::utils;
+using std::string;
+
+namespace {
+    struct MapFind_fixture
```
Comment (pbougue) [inline tests/map_find_test.cpp]: Not fan of the case mix here : I'm not sure, but it's probably more full CamelCase we use in that case.

File: tests/map_find_test.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            map[3] = "three";
+
+            umap[1] = "one";
+            umap[2] = "two";
+            umap[3] = "three";
+        }
+    };
+} // namespace
+
+BOOST_FIXTURE_TEST_CASE(MapFind_should_find_key_in_map, MapFind_fixture) {
```
Comment (pbougue) [inline tests/map_find_test.cpp]: Mixed_case on test name: but not sure what's our agreement here :)

File: map_find.h (hunk 1 above)
Comment (nonifier) [reply map_find.h]: I agree this means exactly the same thing (and generates the exact same code).
The only thing I'm not fond of if we use a pointer, would be the double deref like:
```cpp
MapFind& if_found(std::function<void(const ValueType &)> fn) {
        if( res ) {
            fn(**res);
        }
[...]
```
But if a pointer makes the code easier to read, I'm happy to change it.

File: map_find.h [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+ * @brief Create a MapFind object with the associated map
+ *
+ * This function facilitates the contruction of a MapFind
+ * as it allowes type deduction.
+ * @param map : the map to perform the search on
+ * @param key : the key to search in the map
+ * @returns A MapFind object to query the restult or perfom another find
+ */
+template<class Map, class Key>
+MapFind<Map> make_map_find(Map& c, const Key& key) {
```
Comment (xlqian) [inline map_find.h]: how about add a default value if not found?

Comment (nonifier) [reply map_find.h]: you already have something for this in `functions.h`
```cpp
template<typename Map>
const typename Map::mapped_type&
find_or_default(const typename Map::key_type& k, const Map& m);
```

**Pull request merged**